src/Microsoft.AspNet.Authentication.OAuth/Notifications/OAuthAuthenticatedContext.cs

```diff
@@ -20,55 +22,97 @@ public class OAuthAuthenticatedContext : BaseContext<OAuthAuthenticationOptions>
+        /// <param name="tokens">The tokens returned from the token endpoint.</param>
+        /// <param name="user">The JSON-serialized user.</param>
+        public OAuthAuthenticatedContext(
+            [NotNull] HttpContext context,
+            [NotNull] OAuthAuthenticationOptions options,
+            [NotNull] HttpClient backchannel,
+            [NotNull] OAuthTokenResponse tokens,
+            [NotNull] JObject user)
+            : base(context, options)
+        {
+            TokenResponse = tokens;
+            Backchannel = backchannel;
         }
 
         /// <summary>
-        /// Gets the JSON-serialized user.
+        /// Gets the JSON-serialized user or an empty
+        /// <see cref="JObject"/> if it is not available.
+        /// </summary>
+        public JObject User { get; }
+
+        /// <summary>
+        /// Gets the token response returned by the authentication service.
         /// </summary>
-        public JObject User { get; protected set; }
+        public OAuthTokenResponse TokenResponse { get; }
 
         /// <summary>
         /// Gets the access token provided by the authentication service.
         /// </summary>
-        public string AccessToken { get; protected set; }
+        public string AccessToken => TokenResponse.AccessToken;
 
         /// <summary>
         /// Gets the access token type provided by the authentication service.
         /// </summary>
-        public string TokenType { get; protected set; }
+        public string TokenType => TokenResponse.TokenType;
 
         /// <summary>
         /// Gets the refresh token provided by the authentication service.
         /// </summary>
-        public string RefreshToken { get; protected set; }
+        public string RefreshToken => TokenResponse.RefreshToken;
 
         /// <summary>
         /// Gets the access token expiration time.
         /// </summary>
-        public TimeSpan? ExpiresIn { get; protected set; }
+        public TimeSpan? ExpiresIn
+        {
+            get
+            {
+                int value;
+                if (int.TryParse(TokenResponse.ExpiresIn, NumberStyles.Integer, CultureInfo.InvariantCulture, out value))
+                {
+                    return TimeSpan.FromSeconds(value);
+                }
+
+                return null;
+            }
+        }
+        
+        /// <summary>
+        /// Gets the backchannel used to communicate with the provider.
+        /// </summary>
+        public HttpClient Backchannel { get; }
 
         /// <summary>
-        /// Gets the <see cref="ClaimsIdentity"/> representing the user.
+        /// Gets the <see cref="ClaimsPrincipal"/> representing the user.
         /// </summary>
         public ClaimsPrincipal Principal { get; set; }
 
         /// <summary>
+        /// Gets the main identity exposed by <see cref="Principal"/>.
+        /// </summary>
+        public ClaimsIdentity Identity => (ClaimsIdentity) Principal.Identity;
```

Comment:
@HaoK @Tratcher is it okay to have a new `Identity` property here, that simply wraps `Principal.Identity`?
It makes adding new claims to the principal much easier from the `OnAuthenticated` notification, as you don't have to cast the identity anymore from the user code (see the social sample).